### ISSUE TITLE
Flag inappropriate content

### ISSUE THREAD

**Issue by ExoticObjects** [shyambhat/InstagramKit#100]
This is not a technical issue, but more of a question. The App Store now has these guidelines in place regarding apps that display user-generated content:

> 14.3: Apps that display user generated content must include a method for filtering objectionable material, a mechanism for users to flag offensive content, and the ability to block abusive users from the service .

> It is necessary that you put all of the following precautions in place:

> * Require that users agree to terms (EULA) and these terms must make it clear that there is no tolerance for objectionable content
* Use moderators to flag and remove inappropriate content and offensive users
* Users need a mechanism to flag objectionable content and report users generating this content
* Developer must act on objectionable content reports within 24 hours by removing the content and ejecting the user who provided the offending content
* Developer needs a method for ejecting users who violate the terms of the EULA

As far as I can tell, the Instagram API does not support flagging content. Can you (or anyone) think of a way to achieve this? If they're going to start rejecting apps for this reason, it's not great news for InstagramKit! It would be great to have some sort of workaround...

**Comment by shyambhat** [comment]
Hi @ExoticObjects,
a. Apps using InstagramKit do not 'generate' user content on Instagram. At best, they can pass media to the Instagram iOS App with it's hooks, which makes the Instagram App the creator of content. 
b. Even if it could, as long as the Instagram API does not have endpoints to flag content, there is no way InstagramKit can achieve this.
c. This requirement was introduced by Apple in September 2014. Apps using InstagramKit continue to be approved as of today, April 2015.

**Comment by ExoticObjects** [comment]
I agree with everything you said. But our app was rejected last week. We
made all of the arguments you just made. Nonetheless, they are forcing us
to add a "flag inappropriate content" feature.

It makes no sense, of course. We even lost the appeal.

**Comment by shyambhat** [comment]
@ExoticObjects I looked up more on this and it's hard to imagine Apple enforcing server-side infrastructure for every user content based App out there -- especially for those only displaying external content.

How do you intend to fix this? 

Although I'm afraid InstagramKit can do little to mitigate this, I'll keep the issue open for anyone who's experienced something similar.

**Comment by ExoticObjects** [comment]
We found it hard to imagine as well. But it happened! I wish I was making
it up...

This is not Instagram's or InstagramKit's problem - it's yet another
strange decision from Apple.

We're fixing it by creating a service where users can flag content from
within the app and hide it.

**Comment by dezinezync** [comment]
@ExoticObjects Hey, @shyambhat and I discussed this yesterday. I had ran into a similar issue with Apple, but for another service's API. That API didn't expose flagging content either. So I simply stated that to Apple, and they approved it. You could try that.

**Comment by ExoticObjects** [comment]
I did, of course, but thanks!

**Comment by Rizwan12A** [comment]
HI.....
i ask them that about but they say you have to block these images & videos

**Comment by jacksonh** [comment]
I was also just rejected for this. I've contacted a friend to see about getting flagging added to the API, but that will obviously take a while.

**Comment by StevenGurr** [comment]
I've now been rejected for this as well. I naively thought it'd be quite simple to explain that it's just pulling the data from Instagram, but the responses are phrased like they've never heard of Instagram.

They want me to add an EULA to make it clear there's no tolerance for 'objectional content', employ moderators to flag and remove inappropriate content and users(!), allow users to flag objections content, allow users to block other users, and I've got to be able to 'act on' reports of objectional content within 24 hours by blocking content and removing the user, and be able to eject users who violate the EULA I've got to make up.

Utterly ridiculous. I only made a very simple app for browsing photos. :-(

**Comment by angellee** [comment]
Same here. My app got rejected as well. Apple doesn't care what you said. I explained that the api does not provide a way to flag/remove inappropriate contents (they do on their end but it is not exposed to users) and even cited similar apps which are currently on app store that uses the same api WITHOUT any of the 14.3 items they mentioned. I bet they didn't even read my explanation. Every time they replied, they just re-pasted the 14.3 items and that's it. My app doesn't even allow users to sign up. It is sign in only. How does it make sense to prompt EULA on a sign-in only app? Totally absurd.

**Comment by anclefeudor** [comment]
Same problem. Guys if someone resolve this, please write!

**Comment by Pranoy1c** [comment]
If you are allowing users to upload content:
1. When launching the app, make the user agree to an EULA which can be presented in a textview saying they agree not to post offensive stuff.
2. Have an algorithm which analyzes the pixels of each picture and calculates the probability of something being NSFW and the text comment to detect "Offensive" words like "Fuck" etc.
3. Also mention to apple that when posted through your app, the content goes to instagram and instagram provides a "Report" and Block button within the app.

These resolved my issues. I used the expedited review option to fasten things up.
